src/partition/partitioner.cpp

```diff
@@ -57,13 +60,74 @@ struct CompressedNodeBasedGraph
 CompressedNodeBasedGraph LoadCompressedNodeBasedGraph(const std::string &path)
 {
     const auto fingerprint = storage::io::FileReader::VerifyFingerprint;
-
     storage::io::FileReader reader(path, fingerprint);
 
     CompressedNodeBasedGraph graph{reader};
     return graph;
 }
 
+struct NodeBasedGraphToEdgeBasedGraphMapping
+{
+    NodeBasedGraphToEdgeBasedGraphMapping(storage::io::FileReader &reader)
+    {
+        // Reads:  | Fingerprint | #mappings | u v head tail | u v head tail | ..
+        // - uint64: number of mappings (u, v, head, tail) chunks
+        // - NodeID u, NodeID v, EdgeID head, EdgeID tail
+        //
+        // Gets written in NodeBasedGraphToEdgeBasedGraphMappingWriter
+
+        const auto num_mappings = reader.ReadElementCount64();
+
+        heads.reserve(num_mappings);
+        tails.reserve(num_mappings);
+
+        for (std::uint64_t i{0}; i < num_mappings; ++i)
+        {
+
+            const auto u = reader.ReadOne<NodeID>();    // node based graph `from` nodes
+            const auto v = reader.ReadOne<NodeID>();    // node based graph `to` nodes
+            const auto head = reader.ReadOne<EdgeID>(); // edge based graph forward nodes
+            const auto tail = reader.ReadOne<EdgeID>(); // edge based graph backward nodes
+
+            heads.insert({head, {u, v}});
+            tails.insert({tail, {u, v}});
+        }
+    }
+
+    struct NodeBasedNodes
+    {
+        NodeID u, v;
+    };
+
+    NodeBasedNodes Lookup(EdgeID edge_based_node)
+    {
+        auto heads_it = heads.find(edge_based_node);
+        if (heads_it != end(heads))
+            return heads_it->second;
+
+        auto tails_it = tails.find(edge_based_node);
+        if (tails_it != end(tails))
+            return tails_it->second;
+
+        BOOST_ASSERT(false);
+        return NodeBasedNodes{SPECIAL_NODEID, SPECIAL_NODEID};
+    }
+
+  private:
+    std::unordered_map<EdgeID, NodeBasedNodes> heads;
```

Comment:
The edge_ids are continuous if you merge `heads` and `tails` into one array. If the information about head/tail is needed, saving that in a continuous `std::vector<bool>` should be more efficient.